### PR TITLE
feat: chunk validator MVP for contract code distribution

### PR DESCRIPTION
**Pull request by pugachAG**
Implements handling of the messaged introduced in #12224:
- `PartialEncodedStateWitnessTracker` is modified to track the accessed contracts.
- Message validation and metrics will be added in a separate PR.

### CODE REVIEW

File: chain/client/src/stateless_validation/partial_witness/partial_witness_tracker.rs [hunk 1]
```diff
@@ -77,13 +121,180 @@ impl CacheEntry {
...
+
+    pub fn update(
+        &mut self,
+        update: CacheUpdate,
+    ) -> Option<(DecodePartialWitnessResult, Vec<CodeBytes>)> {
+        match update {
+            CacheUpdate::WitnessPart(partial_witness, encoder) => {
+                self.process_witness_part(partial_witness, encoder);
+            }
+            CacheUpdate::AccessedContractHashes(hashes) => {
```
[the remark this answers is not in the document]
Comment (tayfunelmas) [reply chain/client/src/stateless_validation/partial_witness/partial_witness_tracker.rs:188]: then rename `codes` also to `code_bytes` below for consistency

File: chain/client/src/stateless_validation/partial_witness/partial_witness_tracker.rs [hunk 2]
```diff
@@ -77,13 +121,180 @@ impl CacheEntry {
...
+            AccessedContractsState::Unknown | AccessedContractsState::Received(_)
+        );
+        if !(parts_ready && contracts_ready) {
+            return None;
+        }
+        let decode_result = match &mut self.witness_parts {
+            WitnessPartsState::Empty | WitnessPartsState::WaitingParts(_) => unreachable!(),
+            WitnessPartsState::Decoded(_) => {
+                // We want to avoid copying decoded witness, so we move it out of the state
+                // are reset it to Empty.
```
Comment (tayfunelmas) [inline chain/client/src/stateless_validation/partial_witness/partial_witness_tracker.rs:273]: nit. `are` --> `and`?

File: chain/client/src/stateless_validation/partial_witness/partial_witness_tracker.rs [hunk 3]
```diff
@@ -77,13 +121,180 @@ impl CacheEntry {
...
+                };
+                result
+            }
+        };
+        let contracts: Vec<CodeBytes> = match &mut self.accessed_contracts {
+            AccessedContractsState::Unknown => vec![],
+            AccessedContractsState::Requested(_) => unreachable!(),
+            AccessedContractsState::Received(_) => {
+                // We want to avoid copying contracts, so we move them out of the state
+                // are reset it to Unknown.
```
Comment (tayfunelmas) [inline chain/client/src/stateless_validation/partial_witness/partial_witness_tracker.rs:287]: `are` -> `and`?

**Review by tayfunelmas — APPROVED** [review]
All nits. thanks. LGTM.

File: chain/client/src/stateless_validation/partial_witness/partial_witness_actor.rs [hunk 4]
```diff
@@ -344,42 +333,87 @@ impl PartialWitnessActor {
...
+        }
+        let key = accesses.chunk_production_key();
+        let contract_hashes = accesses.contracts();
+        self.partial_witness_tracker
+            .store_accessed_contract_hashes(key.clone(), contract_hashes.to_vec())?;
+        // TODO(#11099): currently we always request all hashes to test worst case scenario.
+        // Eventually we want to only request ones that are missing from the complied contracts cache.
+        let random_chunk_producer =
+            self.epoch_manager.get_random_chunk_producers_for_shard(&key.epoch_id, key.shard_id)?;
+        let request = ContractCodeRequest::new(key.clone(), contract_hashes.to_vec(), &signer);
```
[the remark this answers is not in the document]
Comment (pugachAG) [reply chain/client/src/stateless_validation/partial_witness/partial_witness_actor.rs:364]: key is a ref and `::new` takes owned instance, so we do need clone here

**Pull request merged**